codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocket08FrameDecoder.java

```diff
@@ -409,19 +418,28 @@ private void unmask(ByteBuf frame) {
     }
 
     private void protocolViolation(ChannelHandlerContext ctx, String reason) {
-        protocolViolation(ctx, new CorruptedFrameException(reason));
+        protocolViolation(ctx, WebSocketCloseStatus.PROTOCOL_ERROR, reason);
+    }
+
+    private void protocolViolation(ChannelHandlerContext ctx, WebSocketCloseStatus status, String reason) {
+        protocolViolation(ctx, new CorruptedWebSocketFrameException(status, reason));
     }
 
-    private void protocolViolation(ChannelHandlerContext ctx, CorruptedFrameException ex) {
+    private void protocolViolation(ChannelHandlerContext ctx, CorruptedWebSocketFrameException ex) {
         state = State.CORRUPT;
-        if (ctx.channel().isActive()) {
+        if (ctx.channel().isActive() && config.closeOnProtocolViolation()) {
             Object closeMessage;
             if (receivedClosingHandshake) {
                 closeMessage = Unpooled.EMPTY_BUFFER;
             } else {
-                closeMessage = new CloseWebSocketFrame(1002, null);
+                WebSocketCloseStatus closeStatus = ex.getCloseStatus();
+                String reasonText = ex.getMessage();
+                if (reasonText == null || reasonText.isEmpty()) {
+                    reasonText = closeStatus.reasonText();
+                }
+                closeMessage = new CloseWebSocketFrame(closeStatus, reasonText);
             }
-            ctx.writeAndFlush(closeMessage).addListener(ChannelFutureListener.CLOSE);
+            ctx.channel().writeAndFlush(closeMessage).addListener(ChannelFutureListener.CLOSE);
```

Comment:
revert this ... `ctx.writeAndFlush(...)` is correct.

Reply:
As I see in 'WebSocketServerHandshaker#handshake' codecs are added in the following order:

    p.addBefore(ctx.name(), "wsdecoder", newWebsocketDecoder());
    p.addBefore(ctx.name(), "wsencoder", newWebSocketEncoder());

this means, when decoder calls 'ctx.writeAndFlush' it will never pass through encoder and will be terminated in 'AbstractNioByteChannel.filterOutboundMessage'.

Reply:
anyway, in our applications we will use 'closeOnProtocolViolation=false', so It would not be a problem. 
reverted.